pd crashed during unsafe remove-failed-stores

## Bug Report

<!-- Please answer these questions before submitting your issue. Thanks! -->

### What did you do?

when execute pd-ctl unsafe remove-failed-stores, the pd crashes with

```
[2022/08/10 16:32:01.142 +00:00] [INFO] [unsafe_recovery_controller.go:505] ["{\"info\":\"Unsafe recovery enters exit force leader stage\",\"time\":\"2022-08-10 16:32:01.142\",\"details\":[\"triggered by error: region \\u0026{0xc06ac4c980 14048}'s epoch shouldn't be staler than old ones \\u0026{0xc0658f71c0 14039}\"]}"]
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1b599d4]

goroutine 834457685 [running]:
github.com/tikv/pd/server/cluster.(*unsafeRecoveryController).generateExitForceLeaderPlan(...)
        /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/pd/server/cluster/unsafe_recovery_controller.go:1154
github.com/tikv/pd/server/cluster.(*unsafeRecoveryController).HandleStoreHeartbeat(0xc01e88bce0, 0x1?, 0x0?)
        /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/pd/server/cluster/unsafe_recovery_controller.go:343 +0x514
github.com/tikv/pd/server.(*GrpcServer).StoreHeartbeat(0xc009d9d760, {0x33acbc0?, 0xc0b8efbad0?}, 0xc0b8eff000)
        /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/pd/server/grpc_service.go:624 +0x5c5
github.com/pingcap/kvproto/pkg/pdpb._PD_StoreHeartbeat_Handler.func1({0x33acbc0, 0xc0b8efbad0}, {0x2685c00?, 0xc0b8eff000})
        /go/pkg/mod/github.com/pingcap/kvproto@v0.0.0-20220510035547-0e2f26c0a46a/pkg/pdpb/pdpb.pb.go:7467 +0x7b
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1({0x33acbc0?, 0xc0b8efbad0?}, {0x2685c00?, 0xc0b8eff000?})
        /go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:31 +0x89
github.com/grpc-ecosystem/go-grpc-prometheus.(*ServerMetrics).UnaryServerInterceptor.func1({0x33acbc0, 0xc0b8efbad0}, {0x2685c00, 0xc0b8eff000}, 0x24ac200?, 0xc0358f6f00)
        /go/pkg/mod/github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:107 +0x87
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1({0x33acbc0?, 0xc0b8efbad0?}, {0x2685c00?, 0xc0b8eff000?})
        /go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:34 +0x6f
go.etcd.io/etcd/etcdserver/api/v3rpc.newUnaryInterceptor.func1({0x33acbc0, 0xc0b8efbad0}, {0x2685c00?, 0xc0b8eff000}, 0x0?, 0xc0358f6f00)
        /go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20191023171146-3cf2f69b5738/etcdserver/api/v3rpc/interceptor.go:64 +0x1b6
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1({0x33acbc0?, 0xc0b8efbad0?}, {0x2685c00?, 0xc0b8eff000?})
        /go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:34 +0x6f
go.etcd.io/etcd/etcdserver/api/v3rpc.newLogUnaryInterceptor.func1({0x33acbc0, 0xc0b8efbad0}, {0x2685c00, 0xc0b8eff000}, 0xc0bcf2a140, 0xc0358f6f00)
        /go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20191023171146-3cf2f69b5738/etcdserver/api/v3rpc/interceptor.go:71 +0xc3
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1({0x33acbc0, 0xc0b8efbad0}, {0x2685c00, 0xc0b8eff000}, 0xc0bcf2a140, 0xc050ccfae8)
        /go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:39 +0x1a3
github.com/pingcap/kvproto/pkg/pdpb._PD_StoreHeartbeat_Handler({0x2700f60?, 0xc009d9d760}, {0x33acbc0, 0xc0b8efbad0}, 0xc08afe4720, 0xc001f8d350)
        /go/pkg/mod/github.com/pingcap/kvproto@v0.0.0-20220510035547-0e2f26c0a46a/pkg/pdpb/pdpb.pb.go:7469 +0x138
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00067e480, {0x33b7950, 0xc0676b1b90}, 0xc028e50c00, 0xc000907d10, 0x433d028, 0x0)
        /go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:1024 +0xd1e
google.golang.org/grpc.(*Server).handleStream(0xc00067e480, {0x33b7950, 0xc0676b1b90}, 0xc028e50c00, 0x0)
        /go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:1313 +0xa05
google.golang.org/grpc.(*Server).serveStreams.func1.1()
        /go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:722 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:720 +0xea
```
however, retry the command could succeed

### What did you expect to see?

the command doesn't crash pd, either succeed or fail gracefully with meaningful error message

### What did you see instead?

pd crashed

### What version of PD are you using (`pd-server -V`)?
v6.1.0

--- a/server/cluster/unsafe_recovery_controller.go
+++ b/server/cluster/unsafe_recovery_controller.go
@@ -360,7 +360,6 @@ func (u *unsafeRecoveryController) HandleStoreHeartbeat(heartbeat *pdpb.StoreHea
 				if hasPlan, err = u.generateCreateEmptyRegionPlan(newestRegionTree, peersMap); hasPlan && err == nil {
 					u.changeStage(createEmptyRegion)
 					break
-
 				}
 				if err != nil {
 					break

--- a/server/cluster/unsafe_recovery_controller.go
+++ b/server/cluster/unsafe_recovery_controller.go
@@ -303,19 +303,12 @@ func (u *unsafeRecoveryController) HandleStoreHeartbeat(heartbeat *pdpb.StoreHea
 			u.changeStage(failed)
 			return
 		} else {
-			// Manually set stage to exit force leader so that we do not have to collect reports twice.
+			// Manually set stage to exit force leader instead of calling changeStage(). So that
+			// 1. TiKV reports are not cleaned up here.
+			// 2. Exit force leader plan can still be genereated from the reports.
+			// 3. Clean up the reports after the plan is genereated by calling changeStage() there.
 			stage = exitForceLeader
-			var output StageOutput
-			output.Time = time.Now().Format("2006-01-02 15:04:05.000")
-			output.Info = "Unsafe recovery enters exit force leader stage"
-			output.Details = append(output.Details, fmt.Sprintf("triggered by error: %v", err.Error()))
-			u.output = append(u.output, output)
-			data, marshalErr := json.Marshal(output)
-			if marshalErr != nil {
-				log.Error("Unsafe recovery fail to marshal json object", zap.String("err", marshalErr.Error()))
-			} else {
-				log.Error(string(data))
-			}
+			u.err = err
 		}
 		reCheck := false
 		for {

--- a/server/cluster/unsafe_recovery_controller.go
+++ b/server/cluster/unsafe_recovery_controller.go
@@ -492,7 +492,6 @@ func (u *unsafeRecoveryController) changeStage(stage unsafeRecoveryStage) {
 		output.Info = "Unsafe recovery enters exit force leader stage"
 		if u.err != nil {
 			output.Details = append(output.Details, fmt.Sprintf("triggered by error: %v", u.err.Error()))
-			u.err = nil
 		}
 	case finished:
 		if u.step > 1 {
@@ -517,7 +516,7 @@ func (u *unsafeRecoveryController) changeStage(stage unsafeRecoveryStage) {
 	u.output = append(u.output, output)
 	data, err := json.Marshal(output)
 	if err != nil {
-		log.Error("Unsafe recovery fail to marshal json object", zap.String("err", err.Error()))
+		log.Error("Unsafe recovery fail to marshal json object", zap.Error(err))
 	} else {
 		log.Info(string(data))
 	}

--- a/server/cluster/unsafe_recovery_controller.go
+++ b/server/cluster/unsafe_recovery_controller.go
@@ -492,6 +492,7 @@ func (u *unsafeRecoveryController) changeStage(stage unsafeRecoveryStage) {
 		output.Info = "Unsafe recovery enters exit force leader stage"
 		if u.err != nil {
 			output.Details = append(output.Details, fmt.Sprintf("triggered by error: %v", u.err.Error()))
+			u.err = nil
 		}
 	case finished:
 		if u.step > 1 {

--- a/server/cluster/unsafe_recovery_controller.go
+++ b/server/cluster/unsafe_recovery_controller.go
@@ -298,7 +298,7 @@ func (u *unsafeRecoveryController) HandleStoreHeartbeat(heartbeat *pdpb.StoreHea
 
 	if allCollected {
 		newestRegionTree, peersMap, buildErr := u.buildUpFromReports()
-		if u.HandleErr(buildErr) {
+		if buildErr != nil && u.HandleErr(buildErr) {
 			return
 		}
 

--- a/server/cluster/unsafe_recovery_controller.go
+++ b/server/cluster/unsafe_recovery_controller.go
@@ -262,7 +262,8 @@ func (u *unsafeRecoveryController) checkTimeout() bool {
 
 	if time.Now().After(u.timeout) {
 		u.err = errors.Errorf("Exceeds timeout %v", u.timeout)
-		return u.handleErr()
+		u.changeStage(failed)
+		return true
 	}
 	return false
 }
@@ -281,23 +282,27 @@ func (u *unsafeRecoveryController) HandleStoreHeartbeat(heartbeat *pdpb.StoreHea
 		return
 	}
 
-	allCollected, err := u.collectReport(heartbeat)
-	if err != nil {
-		u.err = err
-		if u.handleErr() {
-			return
-		}
+	storeID := heartbeat.Stats.StoreId
+	if _, isFailedStore := u.failedStores[storeID]; isFailedStore {
+		u.err = errors.Errorf("Receive heartbeat from failed store %d", storeID)
+		return
 	}
 
+	allCollected := u.collectReport(heartbeat)
+
 	if allCollected {
-		newestRegionTree, peersMap, err := u.buildUpFromReports()
+		newestRegionTree, peersMap, buildErr := u.buildUpFromReports()
+		// Keep the earliest error.
+		if buildErr != nil && u.err == nil {
+			u.err = buildErr
+		}
 
 		// clean up previous plan
 		u.storePlanExpires = make(map[uint64]time.Time)
 		u.storeRecoveryPlans = make(map[uint64]*pdpb.RecoveryPlan)
 
 		var stage unsafeRecoveryStage
-		if err == nil {
+		if u.err == nil {
 			stage = u.stage
 		} else if u.stage == exitForceLeader {
 			u.changeStage(failed)
@@ -308,33 +313,47 @@ func (u *unsafeRecoveryController) HandleStoreHeartbeat(heartbeat *pdpb.StoreHea
 			// 2. Exit force leader plan can still be genereated from the reports.
 			// 3. Clean up the reports after the plan is genereated by calling changeStage() there.
 			stage = exitForceLeader
-			u.err = err
 		}
 		reCheck := false
+		hasPlan := false
+		var err error
 		for {
 			switch stage {
 			case collectReport:
 				fallthrough
 			case tombstoneTiFlashLearner:
-				if u.generateTombstoneTiFlashLearnerPlan(newestRegionTree, peersMap) {
+				hasPlan, err = u.generateTombstoneTiFlashLearnerPlan(newestRegionTree, peersMap)
+				if err != nil {
+					break
+				}
+				if hasPlan {
 					u.changeStage(tombstoneTiFlashLearner)
 					break
 				}
 				fallthrough
 			case forceLeaderForCommitMerge:
-				if u.generateForceLeaderPlan(newestRegionTree, peersMap, true) {
+				hasPlan, err = u.generateForceLeaderPlan(newestRegionTree, peersMap, true)
+				if err != nil {
+					break
+				}
+				if hasPlan {
 					u.changeStage(forceLeaderForCommitMerge)
 					break
 				}
 				fallthrough
 			case forceLeader:
-				if u.generateForceLeaderPlan(newestRegionTree, peersMap, false) {
+				hasPlan, err = u.generateForceLeaderPlan(newestRegionTree, peersMap, false)
+				if err != nil {
+					break
+				}
+				if hasPlan {
 					u.changeStage(forceLeader)
 					break
 				}
 				fallthrough
 			case demoteFailedVoter:
-				if u.generateDemoteFailedVoterPlan(newestRegionTree, peersMap) {
+				hasPlan = u.generateDemoteFailedVoterPlan(newestRegionTree, peersMap)
+				if hasPlan {
 					u.changeStage(demoteFailedVoter)
 					break
 				} else if !reCheck {
@@ -344,27 +363,45 @@ func (u *unsafeRecoveryController) HandleStoreHeartbeat(heartbeat *pdpb.StoreHea
 				}
 				fallthrough
 			case createEmptyRegion:
-				if u.generateCreateEmptyRegionPlan(newestRegionTree, peersMap) {
+				hasPlan, err = u.generateCreateEmptyRegionPlan(newestRegionTree, peersMap)
+				if err != nil {
+					break
+				}
+				if hasPlan {
 					u.changeStage(createEmptyRegion)
 					break
 				}
 				fallthrough
 			case exitForceLeader:
 				// no need to generate plan, empty recovery plan triggers exit force leader on TiKV side
-				if u.generateExitForceLeaderPlan() {
+				hasPlan = u.generateExitForceLeaderPlan()
+				if hasPlan {
 					u.changeStage(exitForceLeader)
 				}
 			default:
 				panic("unreachable")
 			}
 
-			hasPlan := len(u.storeRecoveryPlans) != 0
-			if u.err != nil {
-				if u.handleErr() {
-					return
+			if err != nil {
+				// Keep the earliest error.
+				if u.err != nil {
+					u.err = err
 				}
+				if stage == exitForceLeader {
+					u.changeStage(failed)
+				} else {
+					u.storePlanExpires = make(map[uint64]time.Time)
+					u.storeRecoveryPlans = make(map[uint64]*pdpb.RecoveryPlan)
+					u.timeout = time.Now().Add(storeRequestInterval)
+					u.changeStage(exitForceLeader)
+				}
+				return
 			} else if !hasPlan {
-				u.changeStage(finished)
+				if u.err != nil {
+					u.changeStage(failed)
+				} else {
+					u.changeStage(finished)
+				}
 				return
 			}
 			break
@@ -372,20 +409,6 @@ func (u *unsafeRecoveryController) HandleStoreHeartbeat(heartbeat *pdpb.StoreHea
 	}
 
 	u.dispatchPlan(heartbeat, resp)
-}
-
-func (u *unsafeRecoveryController) handleErr() bool {
-	if u.err != nil {
-		if u.stage == exitForceLeader {
-			u.changeStage(failed)
-			return true
-		}
-		u.storePlanExpires = make(map[uint64]time.Time)
-		u.storeRecoveryPlans = make(map[uint64]*pdpb.RecoveryPlan)
-		u.timeout = time.Now().Add(storeRequestInterval)
-		u.changeStage(exitForceLeader)
-	}
-	return false
 }
 
 // It dispatches recovery plan if any.
@@ -411,21 +434,18 @@ func (u *unsafeRecoveryController) dispatchPlan(heartbeat *pdpb.StoreHeartbeatRe
 }
 
 // It collects and checks if store reports have been fully collected.
-func (u *unsafeRecoveryController) collectReport(heartbeat *pdpb.StoreHeartbeatRequest) (bool, error) {
+func (u *unsafeRecoveryController) collectReport(heartbeat *pdpb.StoreHeartbeatRequest) bool {
 	storeID := heartbeat.Stats.StoreId
-	if _, isFailedStore := u.failedStores[storeID]; isFailedStore {
-		return false, errors.Errorf("Receive heartbeat from failed store %d", storeID)
-	}
 
 	if heartbeat.StoreReport == nil {
-		return false, nil
+		return false
 	}
 
 	if heartbeat.StoreReport.GetStep() != u.step {
 		log.Info("Unsafe recovery receives invalid store report",
 			zap.Uint64("store-id", storeID), zap.Uint64("expected-step", u.step), zap.Uint64("obtained-step", heartbeat.StoreReport.GetStep()))
 		// invalid store report, ignore
-		return false, nil
+		return false
 	}
 
 	if report, exists := u.storeReports[storeID]; exists {
@@ -434,11 +454,11 @@ func (u *unsafeRecoveryController) collectReport(heartbeat *pdpb.StoreHeartbeatR
 		if report == nil {
 			u.numStoresReported++
 			if u.numStoresReported == len(u.storeReports) {
-				return true, nil
+				return true
 			}
 		}
 	}
-	return false, nil
+	return false
 }
 
 // Gets the stage of the current unsafe recovery.
@@ -492,7 +512,6 @@ func (u *unsafeRecoveryController) changeStage(stage unsafeRecoveryStage) {
 		output.Info = "Unsafe recovery enters exit force leader stage"
 		if u.err != nil {
 			output.Details = append(output.Details, fmt.Sprintf("triggered by error: %v", u.err.Error()))
-			u.err = nil
 		}
 	case finished:
 		if u.step > 1 {
@@ -912,18 +931,19 @@ func (u *unsafeRecoveryController) selectLeader(peersMap map[uint64][]*regionIte
 	return leader
 }
 
-func (u *unsafeRecoveryController) generateTombstoneTiFlashLearnerPlan(newestRegionTree *regionTree, peersMap map[uint64][]*regionItem) bool {
+func (u *unsafeRecoveryController) generateTombstoneTiFlashLearnerPlan(newestRegionTree *regionTree, peersMap map[uint64][]*regionItem) (bool, error) {
 	if u.err != nil {
-		return false
+		return false, nil
 	}
 	hasPlan := false
 
+	var err error
 	newestRegionTree.tree.Ascend(func(item btree.Item) bool {
 		region := item.(*regionItem).Region()
 		if !u.canElectLeader(region, false) {
 			leader := u.selectLeader(peersMap, region)
 			if leader == nil {
-				u.err = errors.Errorf("can't select leader for region %d: %v", region.GetId(), logutil.RedactStringer(core.RegionToHexMeta(region)))
+				err = errors.Errorf("can't select leader for region %d: %v", region.GetId(), logutil.RedactStringer(core.RegionToHexMeta(region)))
 				return false
 			}
 			storeID := leader.storeID
@@ -937,12 +957,12 @@ func (u *unsafeRecoveryController) generateTombstoneTiFlashLearnerPlan(newestReg
 		}
 		return true
 	})
-	return hasPlan
+	return hasPlan, err
 }
 
-func (u *unsafeRecoveryController) generateForceLeaderPlan(newestRegionTree *regionTree, peersMap map[uint64][]*regionItem, forCommitMerge bool) bool {
+func (u *unsafeRecoveryController) generateForceLeaderPlan(newestRegionTree *regionTree, peersMap map[uint64][]*regionItem, forCommitMerge bool) (bool, error) {
 	if u.err != nil {
-		return false
+		return false, nil
 	}
 	hasPlan := false
 
@@ -955,6 +975,7 @@ func (u *unsafeRecoveryController) generateForceLeaderPlan(newestRegionTree *reg
 		return false
 	}
 
+	var err error
 	// Check the regions in newest Region Tree to see if it can still elect leader
 	// considering the failed stores
 	newestRegionTree.tree.Ascend(func(item btree.Item) bool {
@@ -971,14 +992,14 @@ func (u *unsafeRecoveryController) generateForceLeaderPlan(newestRegionTree *reg
 				// propose an empty raft log on being leader
 				return true
 			} else if !forCommitMerge && report.HasCommitMerge {
-				u.err = errors.Errorf("unexpected commit merge state for report %v", report)
+				err = errors.Errorf("unexpected commit merge state for report %v", report)
 				return false
 			}
 			// the peer with largest log index/term may have lower commit/apply index, namely, lower epoch version
 			// so find which peer should to be the leader instead of using peer info in the region tree.
 			leader := u.selectLeader(peersMap, region)
 			if leader == nil {
-				u.err = errors.Errorf("can't select leader for region %d: %v", region.GetId(), logutil.RedactStringer(core.RegionToHexMeta(region)))
+				err = errors.Errorf("can't select leader for region %d: %v", region.GetId(), logutil.RedactStringer(core.RegionToHexMeta(region)))
 				return false
 			}
 			storeRecoveryPlan := u.getRecoveryPlan(leader.storeID)
@@ -1006,7 +1027,7 @@ func (u *unsafeRecoveryController) generateForceLeaderPlan(newestRegionTree *reg
 		}
 	}
 
-	return hasPlan
+	return hasPlan, err
 }
 
 func (u *unsafeRecoveryController) generateDemoteFailedVoterPlan(newestRegionTree *regionTree, peersMap map[uint64][]*regionItem) bool {
@@ -1067,9 +1088,9 @@ func (u *unsafeRecoveryController) generateDemoteFailedVoterPlan(newestRegionTre
 	return hasPlan
 }
 
-func (u *unsafeRecoveryController) generateCreateEmptyRegionPlan(newestRegionTree *regionTree, peersMap map[uint64][]*regionItem) bool {
+func (u *unsafeRecoveryController) generateCreateEmptyRegionPlan(newestRegionTree *regionTree, peersMap map[uint64][]*regionItem) (bool, error) {
 	if u.err != nil {
-		return false
+		return false, nil
 	}
 	hasPlan := false
 
@@ -1100,6 +1121,7 @@ func (u *unsafeRecoveryController) generateCreateEmptyRegionPlan(newestRegionTre
 		return 0
 	}
 
+	var err error
 	// There may be ranges that are covered by no one. Find these empty ranges, create new
 	// regions that cover them and evenly distribute newly created regions among all stores.
 	lastEnd := []byte("")
@@ -1112,13 +1134,13 @@ func (u *unsafeRecoveryController) generateCreateEmptyRegionPlan(newestRegionTre
 				storeID = getRandomStoreID()
 				// can't create new region on tiflash store, choose a random one
 				if storeID == 0 {
-					u.err = errors.New("can't find available store(exclude tiflash) to create new region")
+					err = errors.New("can't find available store(exclude tiflash) to create new region")
 					return false
 				}
 			}
-			newRegion, err := createRegion(lastEnd, region.StartKey, storeID)
-			if err != nil {
-				u.err = err
+			newRegion, createRegionErr := createRegion(lastEnd, region.StartKey, storeID)
+			if createRegionErr != nil {
+				err = createRegionErr
 				return false
 			}
 			// paranoid check: shouldn't overlap with any of the peers
@@ -1131,7 +1153,7 @@ func (u *unsafeRecoveryController) generateCreateEmptyRegionPlan(newestRegionTre
 						(len(newRegion.EndKey) == 0 || bytes.Compare(peer.Region().StartKey, newRegion.EndKey) < 0)) ||
 						((len(peer.Region().EndKey) == 0 || bytes.Compare(newRegion.StartKey, peer.Region().EndKey) < 0) &&
 							(len(newRegion.EndKey) == 0 || (len(peer.Region().EndKey) != 0 && bytes.Compare(peer.Region().EndKey, newRegion.EndKey) <= 0))) {
-						u.err = errors.Errorf(
+						err = errors.Errorf(
 							"Find overlap peer %v with newly created empty region %v",
 							logutil.RedactStringer(core.RegionToHexMeta(peer.Region())),
 							logutil.RedactStringer(core.RegionToHexMeta(newRegion)),
@@ -1149,8 +1171,8 @@ func (u *unsafeRecoveryController) generateCreateEmptyRegionPlan(newestRegionTre
 		lastStoreID = storeID
 		return true
 	})
-	if u.err != nil {
-		return false
+	if err != nil {
+		return false, err
 	}
 
 	if !bytes.Equal(lastEnd, []byte("")) || newestRegionTree.size() == 0 {
@@ -1158,21 +1180,19 @@ func (u *unsafeRecoveryController) generateCreateEmptyRegionPlan(newestRegionTre
 			// the last store id is invalid, so choose a random one
 			lastStoreID = getRandomStoreID()
 			if lastStoreID == 0 {
-				u.err = errors.New("can't find available store(exclude tiflash) to create new region")
-				return false
+				return false, errors.New("can't find available store(exclude tiflash) to create new region")
 			}
 		}
 		newRegion, err := createRegion(lastEnd, []byte(""), lastStoreID)
 		if err != nil {
-			u.err = err
-			return false
+			return false, err
 		}
 		storeRecoveryPlan := u.getRecoveryPlan(lastStoreID)
 		storeRecoveryPlan.Creates = append(storeRecoveryPlan.Creates, newRegion)
 		u.recordAffectedRegion(newRegion)
 		hasPlan = true
 	}
-	return hasPlan
+	return hasPlan, nil
 }
 
 func (u *unsafeRecoveryController) generateExitForceLeaderPlan() bool {

--- a/server/cluster/unsafe_recovery_controller.go
+++ b/server/cluster/unsafe_recovery_controller.go
@@ -445,6 +445,7 @@ func (u *unsafeRecoveryController) changeStage(stage unsafeRecoveryStage) {
 	u.stage = stage
 
 	var output StageOutput
+	exitForceLeaderDueToFailure := false
 	output.Time = time.Now().Format("2006-01-02 15:04:05.000")
 	switch u.stage {
 	case idle:
@@ -485,6 +486,7 @@ func (u *unsafeRecoveryController) changeStage(stage unsafeRecoveryStage) {
 		output.Info = "Unsafe recovery enters exit force leader stage"
 		if u.err != nil {
 			output.Details = append(output.Details, fmt.Sprintf("triggered by error: %v", u.err.Error()))
+			exitForceLeaderDueToFailure = true
 		}
 	case finished:
 		if u.step > 1 {
@@ -516,8 +518,10 @@ func (u *unsafeRecoveryController) changeStage(stage unsafeRecoveryStage) {
 
 	// reset store reports to nil instead of delete, because it relays on the item
 	// to decide which store it needs to collect the report from.
-	for k := range u.storeReports {
-		u.storeReports[k] = nil
+	if !exitForceLeaderDueToFailure {
+		for k := range u.storeReports {
+			u.storeReports[k] = nil
+		}
 	}
 	u.numStoresReported = 0
 	u.step += 1
@@ -713,7 +717,7 @@ func (r *regionItem) IsInitialized() bool {
 func (r *regionItem) IsEpochStale(other *regionItem) bool {
 	re := r.Region().GetRegionEpoch()
 	oe := other.Region().GetRegionEpoch()
-	return re.GetVersion() < oe.GetVersion() || re.GetConfVer() < oe.GetConfVer()
+	return re.GetVersion() < oe.GetVersion() || (re.GetVersion() == oe.GetVersion() && re.GetConfVer() < oe.GetConfVer())
 }
 
 func (r *regionItem) IsRaftStale(origin *regionItem, u *unsafeRecoveryController) bool {

--- a/server/cluster/unsafe_recovery_controller.go
+++ b/server/cluster/unsafe_recovery_controller.go
@@ -262,7 +262,7 @@ func (u *unsafeRecoveryController) checkTimeout() bool {
 
 	if time.Now().After(u.timeout) {
 		ret := u.HandleErr(errors.Errorf("Exceeds timeout %v", u.timeout))
-		time.Now().Add(storeRequestInterval)
+		u.timeout = time.Now().Add(storeRequestInterval * 2)
 		return ret
 	}
 	return false

--- a/server/cluster/unsafe_recovery_controller.go
+++ b/server/cluster/unsafe_recovery_controller.go
@@ -291,18 +291,32 @@ func (u *unsafeRecoveryController) HandleStoreHeartbeat(heartbeat *pdpb.StoreHea
 
 	if allCollected {
 		newestRegionTree, peersMap, err := u.buildUpFromReports()
-		if err != nil {
-			u.err = err
-			if u.handleErr() {
-				return
-			}
-		}
 
 		// clean up previous plan
 		u.storePlanExpires = make(map[uint64]time.Time)
 		u.storeRecoveryPlans = make(map[uint64]*pdpb.RecoveryPlan)
 
-		stage := u.stage
+		var stage unsafeRecoveryStage
+		if err == nil {
+			stage = u.stage
+		} else if u.stage == exitForceLeader {
+			u.changeStage(failed)
+			return
+		} else {
+			// Manually set stage to exit force leader so that we do not have to collect reports twice.
+			stage = exitForceLeader
+			var output StageOutput
+			output.Time = time.Now().Format("2006-01-02 15:04:05.000")
+			output.Info = "Unsafe recovery enters exit force leader stage"
+			output.Details = append(output.Details, fmt.Sprintf("triggered by error: %v", err.Error()))
+			u.output = append(u.output, output)
+			data, marshalErr := json.Marshal(output)
+			if marshalErr != nil {
+				log.Error("Unsafe recovery fail to marshal json object", zap.String("err", marshalErr.Error()))
+			} else {
+				log.Error(string(data))
+			}
+		}
 		reCheck := false
 		for {
 			switch stage {
@@ -445,7 +459,6 @@ func (u *unsafeRecoveryController) changeStage(stage unsafeRecoveryStage) {
 	u.stage = stage
 
 	var output StageOutput
-	exitForceLeaderDueToFailure := false
 	output.Time = time.Now().Format("2006-01-02 15:04:05.000")
 	switch u.stage {
 	case idle:
@@ -486,7 +499,6 @@ func (u *unsafeRecoveryController) changeStage(stage unsafeRecoveryStage) {
 		output.Info = "Unsafe recovery enters exit force leader stage"
 		if u.err != nil {
 			output.Details = append(output.Details, fmt.Sprintf("triggered by error: %v", u.err.Error()))
-			exitForceLeaderDueToFailure = true
 		}
 	case finished:
 		if u.step > 1 {
@@ -518,10 +530,8 @@ func (u *unsafeRecoveryController) changeStage(stage unsafeRecoveryStage) {
 
 	// reset store reports to nil instead of delete, because it relays on the item
 	// to decide which store it needs to collect the report from.
-	if !exitForceLeaderDueToFailure {
-		for k := range u.storeReports {
-			u.storeReports[k] = nil
-		}
+	for k := range u.storeReports {
+		u.storeReports[k] = nil
 	}
 	u.numStoresReported = 0
 	u.step += 1

--- a/server/cluster/unsafe_recovery_controller.go
+++ b/server/cluster/unsafe_recovery_controller.go
@@ -261,7 +261,17 @@ func (u *unsafeRecoveryController) checkTimeout() bool {
 	}
 
 	if time.Now().After(u.timeout) {
-		u.err = errors.Errorf("Exceeds timeout %v", u.timeout)
+		return u.HandleErr(errors.Errorf("Exceeds timeout %v", u.timeout))
+	}
+	return false
+}
+
+func (u *unsafeRecoveryController) HandleErr(err error) bool {
+	// Keep the earliest error.
+	if u.err == nil {
+		u.err = err
+	}
+	if u.stage == exitForceLeader {
 		u.changeStage(failed)
 		return true
 	}
@@ -282,19 +292,12 @@ func (u *unsafeRecoveryController) HandleStoreHeartbeat(heartbeat *pdpb.StoreHea
 		return
 	}
 
-	storeID := heartbeat.Stats.StoreId
-	if _, isFailedStore := u.failedStores[storeID]; isFailedStore {
-		u.err = errors.Errorf("Receive heartbeat from failed store %d", storeID)
-		return
-	}
-
 	allCollected := u.collectReport(heartbeat)
 
 	if allCollected {
 		newestRegionTree, peersMap, buildErr := u.buildUpFromReports()
-		// Keep the earliest error.
-		if buildErr != nil && u.err == nil {
-			u.err = buildErr
+		if u.HandleErr(buildErr) {
+			return
 		}
 
 		// clean up previous plan
@@ -304,14 +307,7 @@ func (u *unsafeRecoveryController) HandleStoreHeartbeat(heartbeat *pdpb.StoreHea
 		var stage unsafeRecoveryStage
 		if u.err == nil {
 			stage = u.stage
-		} else if u.stage == exitForceLeader {
-			u.changeStage(failed)
-			return
 		} else {
-			// Manually set stage to exit force leader instead of calling changeStage(). So that
-			// 1. TiKV reports are not cleaned up here.
-			// 2. Exit force leader plan can still be genereated from the reports.
-			// 3. Clean up the reports after the plan is genereated by calling changeStage() there.
 			stage = exitForceLeader
 		}
 		reCheck := false
@@ -383,18 +379,13 @@ func (u *unsafeRecoveryController) HandleStoreHeartbeat(heartbeat *pdpb.StoreHea
 			}
 
 			if err != nil {
-				// Keep the earliest error.
-				if u.err != nil {
-					u.err = err
+				if u.HandleErr(err) {
+					return
 				}
-				if stage == exitForceLeader {
-					u.changeStage(failed)
-				} else {
-					u.storePlanExpires = make(map[uint64]time.Time)
-					u.storeRecoveryPlans = make(map[uint64]*pdpb.RecoveryPlan)
-					u.timeout = time.Now().Add(storeRequestInterval)
-					u.changeStage(exitForceLeader)
-				}
+				u.storePlanExpires = make(map[uint64]time.Time)
+				u.storeRecoveryPlans = make(map[uint64]*pdpb.RecoveryPlan)
+				// Clear the reports etc.
+				u.changeStage(exitForceLeader)
 				return
 			} else if !hasPlan {
 				if u.err != nil {
@@ -436,6 +427,10 @@ func (u *unsafeRecoveryController) dispatchPlan(heartbeat *pdpb.StoreHeartbeatRe
 // It collects and checks if store reports have been fully collected.
 func (u *unsafeRecoveryController) collectReport(heartbeat *pdpb.StoreHeartbeatRequest) bool {
 	storeID := heartbeat.Stats.StoreId
+	if _, isFailedStore := u.failedStores[storeID]; isFailedStore {
+		u.HandleErr(errors.Errorf("Receive heartbeat from failed store %d", storeID))
+		return false
+	}
 
 	if heartbeat.StoreReport == nil {
 		return false

--- a/server/cluster/unsafe_recovery_controller.go
+++ b/server/cluster/unsafe_recovery_controller.go
@@ -499,6 +499,7 @@ func (u *unsafeRecoveryController) changeStage(stage unsafeRecoveryStage) {
 		output.Info = "Unsafe recovery enters exit force leader stage"
 		if u.err != nil {
 			output.Details = append(output.Details, fmt.Sprintf("triggered by error: %v", u.err.Error()))
+			u.err = nil
 		}
 	case finished:
 		if u.step > 1 {
@@ -523,10 +524,10 @@ func (u *unsafeRecoveryController) changeStage(stage unsafeRecoveryStage) {
 	u.output = append(u.output, output)
 	data, err := json.Marshal(output)
 	if err != nil {
-		u.err = err
-		return
+		log.Error("Unsafe recovery fail to marshal json object", zap.String("err", err.Error()))
+	} else {
+		log.Info(string(data))
 	}
-	log.Info(string(data))
 
 	// reset store reports to nil instead of delete, because it relays on the item
 	// to decide which store it needs to collect the report from.

--- a/server/cluster/unsafe_recovery_controller.go
+++ b/server/cluster/unsafe_recovery_controller.go
@@ -261,7 +261,9 @@ func (u *unsafeRecoveryController) checkTimeout() bool {
 	}
 
 	if time.Now().After(u.timeout) {
-		return u.HandleErr(errors.Errorf("Exceeds timeout %v", u.timeout))
+		ret := u.HandleErr(errors.Errorf("Exceeds timeout %v", u.timeout))
+		time.Now().Add(storeRequestInterval)
+		return ret
 	}
 	return false
 }
@@ -318,38 +320,34 @@ func (u *unsafeRecoveryController) HandleStoreHeartbeat(heartbeat *pdpb.StoreHea
 			case collectReport:
 				fallthrough
 			case tombstoneTiFlashLearner:
-				hasPlan, err = u.generateTombstoneTiFlashLearnerPlan(newestRegionTree, peersMap)
-				if err != nil {
+				if hasPlan, err = u.generateTombstoneTiFlashLearnerPlan(newestRegionTree, peersMap); hasPlan && err == nil {
+					u.changeStage(tombstoneTiFlashLearner)
 					break
 				}
-				if hasPlan {
-					u.changeStage(tombstoneTiFlashLearner)
+				if err != nil {
 					break
 				}
 				fallthrough
 			case forceLeaderForCommitMerge:
-				hasPlan, err = u.generateForceLeaderPlan(newestRegionTree, peersMap, true)
-				if err != nil {
+				if hasPlan, err = u.generateForceLeaderPlan(newestRegionTree, peersMap, true); hasPlan && err == nil {
+					u.changeStage(forceLeaderForCommitMerge)
 					break
 				}
-				if hasPlan {
-					u.changeStage(forceLeaderForCommitMerge)
+				if err != nil {
 					break
 				}
 				fallthrough
 			case forceLeader:
-				hasPlan, err = u.generateForceLeaderPlan(newestRegionTree, peersMap, false)
-				if err != nil {
+				if hasPlan, err = u.generateForceLeaderPlan(newestRegionTree, peersMap, false); hasPlan && err == nil {
+					u.changeStage(forceLeader)
 					break
 				}
-				if hasPlan {
-					u.changeStage(forceLeader)
+				if err != nil {
 					break
 				}
 				fallthrough
 			case demoteFailedVoter:
-				hasPlan = u.generateDemoteFailedVoterPlan(newestRegionTree, peersMap)
-				if hasPlan {
+				if hasPlan = u.generateDemoteFailedVoterPlan(newestRegionTree, peersMap); hasPlan {
 					u.changeStage(demoteFailedVoter)
 					break
 				} else if !reCheck {
@@ -359,19 +357,18 @@ func (u *unsafeRecoveryController) HandleStoreHeartbeat(heartbeat *pdpb.StoreHea
 				}
 				fallthrough
 			case createEmptyRegion:
-				hasPlan, err = u.generateCreateEmptyRegionPlan(newestRegionTree, peersMap)
-				if err != nil {
-					break
-				}
-				if hasPlan {
+				if hasPlan, err = u.generateCreateEmptyRegionPlan(newestRegionTree, peersMap); hasPlan && err == nil {
 					u.changeStage(createEmptyRegion)
+					break
+
+				}
+				if err != nil {
 					break
 				}
 				fallthrough
 			case exitForceLeader:
 				// no need to generate plan, empty recovery plan triggers exit force leader on TiKV side
-				hasPlan = u.generateExitForceLeaderPlan()
-				if hasPlan {
+				if hasPlan = u.generateExitForceLeaderPlan(); hasPlan {
 					u.changeStage(exitForceLeader)
 				}
 			default:

--- a/server/cluster/unsafe_recovery_controller_test.go
+++ b/server/cluster/unsafe_recovery_controller_test.go
@@ -1660,7 +1660,7 @@ func TestEpochComparsion(t *testing.T) {
 	cluster := newTestRaftCluster(ctx, mockid.NewIDAllocator(), opt, storage.NewStorageWithMemoryBackend(), core.NewBasicCluster())
 	cluster.coordinator = newCoordinator(ctx, cluster, hbstream.NewTestHeartbeatStreams(ctx, cluster.meta.GetId(), cluster, true))
 	cluster.coordinator.run()
-	for _, store := range newTestStores(4, "6.0.0") {
+	for _, store := range newTestStores(3, "6.0.0") {
 		re.Nil(cluster.PutStore(store.GetMeta()))
 	}
 	recoveryController := newUnsafeRecoveryController(cluster)
@@ -1681,8 +1681,7 @@ func TestEpochComparsion(t *testing.T) {
 						RegionEpoch: &metapb.RegionEpoch{ConfVer: 10, Version: 1},
 						Peers: []*metapb.Peer{
 							{Id: 11, StoreId: 1}, {Id: 21, StoreId: 2}, {Id: 31, StoreId: 3}}}}},
-		}},
-		4: {PeerReports: []*pdpb.PeerReport{
+
 			{
 				RaftState: &raft_serverpb.RaftLocalState{LastIndex: 10, HardState: &eraftpb.HardState{Term: 1, Commit: 10}},
 				RegionState: &raft_serverpb.RegionLocalState{
@@ -1692,53 +1691,41 @@ func TestEpochComparsion(t *testing.T) {
 						EndKey:      []byte(""),
 						RegionEpoch: &metapb.RegionEpoch{ConfVer: 1, Version: 10},
 						Peers: []*metapb.Peer{
-							{Id: 22, StoreId: 2}, {Id: 32, StoreId: 3}, {Id: 42, StoreId: 4}}}}},
+							{Id: 12, StoreId: 1}, {Id: 22, StoreId: 2}, {Id: 32, StoreId: 3}}}}},
 		}},
 	}
 
-	req := newStoreHeartbeat(1, reports[1])
-	req.StoreReport.Step = 1
-	resp := &pdpb.StoreHeartbeatResponse{}
-	recoveryController.HandleStoreHeartbeat(req, resp)
+	advanceUntilFinished(re, recoveryController, reports)
+	expects := map[uint64]*pdpb.StoreReport{
+		1: {PeerReports: []*pdpb.PeerReport{
+			{
+				RaftState: &raft_serverpb.RaftLocalState{LastIndex: 10, HardState: &eraftpb.HardState{Term: 1, Commit: 10}},
+				RegionState: &raft_serverpb.RegionLocalState{
+					Region: &metapb.Region{
+						Id:          1002,
+						StartKey:    []byte("a"),
+						EndKey:      []byte(""),
+						RegionEpoch: &metapb.RegionEpoch{ConfVer: 2, Version: 10},
+						Peers: []*metapb.Peer{
+							{Id: 12, StoreId: 1}, {Id: 22, StoreId: 2, Role: metapb.PeerRole_Learner}, {Id: 32, StoreId: 3, Role: metapb.PeerRole_Learner}}}}},
 
-	req = newStoreHeartbeat(4, reports[4])
-	req.StoreReport.Step = 1
-	resp = &pdpb.StoreHeartbeatResponse{}
-	recoveryController.HandleStoreHeartbeat(req, resp)
-	// enter force leader stage
-	re.Equal(recoveryController.GetStage(), forceLeader)
-	re.NotNil(resp.RecoveryPlan)
-	re.NotNil(resp.RecoveryPlan.ForceLeader)
-	re.Equal(len(resp.RecoveryPlan.ForceLeader.EnterForceLeaders), 1)
-	re.NotNil(resp.RecoveryPlan.ForceLeader.FailedStores)
-	applyRecoveryPlan(re, 4, reports, resp)
-
-	req = newStoreHeartbeat(1, reports[1])
-	resp = &pdpb.StoreHeartbeatResponse{}
-	recoveryController.HandleStoreHeartbeat(req, resp)
-	re.NotNil(resp.RecoveryPlan)
-	re.NotNil(resp.RecoveryPlan.ForceLeader)
-	re.Equal(len(resp.RecoveryPlan.ForceLeader.EnterForceLeaders), 0)
-	applyRecoveryPlan(re, 1, reports, resp)
-
-	// receiving heartbeats
-	req = newStoreHeartbeat(4, reports[4])
-	resp = &pdpb.StoreHeartbeatResponse{}
-	recoveryController.HandleStoreHeartbeat(req, resp)
-	re.Nil(resp.RecoveryPlan)
-	req = newStoreHeartbeat(1, reports[1])
-	resp = &pdpb.StoreHeartbeatResponse{}
-	recoveryController.HandleStoreHeartbeat(req, resp)
-	// enter demote failed voter stage
-	re.Equal(recoveryController.GetStage(), demoteFailedVoter)
-	re.NotNil(resp.RecoveryPlan)
-	re.Equal(len(resp.RecoveryPlan.Tombstones), 1)
-	applyRecoveryPlan(re, 1, reports, resp)
-
-	req = newStoreHeartbeat(4, reports[4])
-	resp = &pdpb.StoreHeartbeatResponse{}
-	recoveryController.HandleStoreHeartbeat(req, resp)
-	re.NotNil(resp.RecoveryPlan)
-	re.Equal(len(resp.RecoveryPlan.Demotes), 1)
-	applyRecoveryPlan(re, 1, reports, resp)
+			{
+				RaftState: &raft_serverpb.RaftLocalState{LastIndex: 10, HardState: &eraftpb.HardState{Term: 1, Commit: 10}},
+				RegionState: &raft_serverpb.RegionLocalState{
+					Region: &metapb.Region{
+						Id:          1,
+						StartKey:    []byte(""),
+						EndKey:      []byte("a"),
+						RegionEpoch: &metapb.RegionEpoch{ConfVer: 1, Version: 1},
+						Peers: []*metapb.Peer{
+							{Id: 2, StoreId: 1}}}}},
+		}},
+	}
+	for storeID, report := range reports {
+		if expect, ok := expects[storeID]; ok {
+			re.Equal(expect.PeerReports, report.PeerReports)
+		} else {
+			re.Empty(len(report.PeerReports))
+		}
+	}
 }

--- a/server/cluster/unsafe_recovery_controller_test.go
+++ b/server/cluster/unsafe_recovery_controller_test.go
@@ -306,6 +306,7 @@ func TestFailed(t *testing.T) {
 		re.NotNil(resp.RecoveryPlan)
 		applyRecoveryPlan(re, storeID, reports, resp)
 	}
+	re.Equal(exitForceLeader, recoveryController.GetStage())
 
 	for storeID, report := range reports {
 		req := newStoreHeartbeat(storeID, report)
@@ -314,7 +315,7 @@ func TestFailed(t *testing.T) {
 		recoveryController.HandleStoreHeartbeat(req, resp)
 		applyRecoveryPlan(re, storeID, reports, resp)
 	}
-	re.Equal(failed, recoveryController.GetStage())
+	re.Equal(finished, recoveryController.GetStage())
 }
 
 func TestForceLeaderFail(t *testing.T) {

--- a/server/cluster/unsafe_recovery_controller_test.go
+++ b/server/cluster/unsafe_recovery_controller_test.go
@@ -296,26 +296,15 @@ func TestFailed(t *testing.T) {
 	resp := &pdpb.StoreHeartbeatResponse{}
 	recoveryController.HandleStoreHeartbeat(req, resp)
 	re.Nil(resp.RecoveryPlan)
-	re.Equal(exitForceLeader, recoveryController.GetStage())
 
 	for storeID, report := range reports {
 		req := newStoreHeartbeat(storeID, report)
 		req.StoreReport = report
 		resp := &pdpb.StoreHeartbeatResponse{}
 		recoveryController.HandleStoreHeartbeat(req, resp)
-		re.NotNil(resp.RecoveryPlan)
-		applyRecoveryPlan(re, storeID, reports, resp)
+		re.Nil(resp.RecoveryPlan)
 	}
-	re.Equal(exitForceLeader, recoveryController.GetStage())
-
-	for storeID, report := range reports {
-		req := newStoreHeartbeat(storeID, report)
-		req.StoreReport = report
-		resp := &pdpb.StoreHeartbeatResponse{}
-		recoveryController.HandleStoreHeartbeat(req, resp)
-		applyRecoveryPlan(re, storeID, reports, resp)
-	}
-	re.Equal(finished, recoveryController.GetStage())
+	re.Equal(failed, recoveryController.GetStage())
 }
 
 func TestForceLeaderFail(t *testing.T) {

--- a/server/cluster/unsafe_recovery_controller_test.go
+++ b/server/cluster/unsafe_recovery_controller_test.go
@@ -1650,3 +1650,95 @@ func TestSplitPaused(t *testing.T) {
 	_, err = cluster.HandleAskBatchSplit(askBatchSplitReq)
 	re.Equal("[PD:unsaferecovery:ErrUnsafeRecoveryIsRunning]unsafe recovery is running", err.Error())
 }
+
+func TestEpochComparsion(t *testing.T) {
+	re := require.New(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	_, opt, _ := newTestScheduleConfig()
+	cluster := newTestRaftCluster(ctx, mockid.NewIDAllocator(), opt, storage.NewStorageWithMemoryBackend(), core.NewBasicCluster())
+	cluster.coordinator = newCoordinator(ctx, cluster, hbstream.NewTestHeartbeatStreams(ctx, cluster.meta.GetId(), cluster, true))
+	cluster.coordinator.run()
+	for _, store := range newTestStores(4, "6.0.0") {
+		re.Nil(cluster.PutStore(store.GetMeta()))
+	}
+	recoveryController := newUnsafeRecoveryController(cluster)
+	re.Nil(recoveryController.RemoveFailedStores(map[uint64]struct{}{
+		2: {},
+		3: {},
+	}, 60, false))
+
+	reports := map[uint64]*pdpb.StoreReport{
+		1: {PeerReports: []*pdpb.PeerReport{
+			{
+				RaftState: &raft_serverpb.RaftLocalState{LastIndex: 10, HardState: &eraftpb.HardState{Term: 1, Commit: 10}},
+				RegionState: &raft_serverpb.RegionLocalState{
+					Region: &metapb.Region{
+						Id:          1001,
+						StartKey:    []byte(""),
+						EndKey:      []byte("b"),
+						RegionEpoch: &metapb.RegionEpoch{ConfVer: 10, Version: 1},
+						Peers: []*metapb.Peer{
+							{Id: 11, StoreId: 1}, {Id: 21, StoreId: 2}, {Id: 31, StoreId: 3}}}}},
+		}},
+		4: {PeerReports: []*pdpb.PeerReport{
+			{
+				RaftState: &raft_serverpb.RaftLocalState{LastIndex: 10, HardState: &eraftpb.HardState{Term: 1, Commit: 10}},
+				RegionState: &raft_serverpb.RegionLocalState{
+					Region: &metapb.Region{
+						Id:          1002,
+						StartKey:    []byte("a"),
+						EndKey:      []byte(""),
+						RegionEpoch: &metapb.RegionEpoch{ConfVer: 1, Version: 10},
+						Peers: []*metapb.Peer{
+							{Id: 22, StoreId: 2}, {Id: 32, StoreId: 3}, {Id: 42, StoreId: 4}}}}},
+		}},
+	}
+
+	req := newStoreHeartbeat(1, reports[1])
+	req.StoreReport.Step = 1
+	resp := &pdpb.StoreHeartbeatResponse{}
+	recoveryController.HandleStoreHeartbeat(req, resp)
+
+	req = newStoreHeartbeat(4, reports[4])
+	req.StoreReport.Step = 1
+	resp = &pdpb.StoreHeartbeatResponse{}
+	recoveryController.HandleStoreHeartbeat(req, resp)
+	// enter force leader stage
+	re.Equal(recoveryController.GetStage(), forceLeader)
+	re.NotNil(resp.RecoveryPlan)
+	re.NotNil(resp.RecoveryPlan.ForceLeader)
+	re.Equal(len(resp.RecoveryPlan.ForceLeader.EnterForceLeaders), 1)
+	re.NotNil(resp.RecoveryPlan.ForceLeader.FailedStores)
+	applyRecoveryPlan(re, 4, reports, resp)
+
+	req = newStoreHeartbeat(1, reports[1])
+	resp = &pdpb.StoreHeartbeatResponse{}
+	recoveryController.HandleStoreHeartbeat(req, resp)
+	re.NotNil(resp.RecoveryPlan)
+	re.NotNil(resp.RecoveryPlan.ForceLeader)
+	re.Equal(len(resp.RecoveryPlan.ForceLeader.EnterForceLeaders), 0)
+	applyRecoveryPlan(re, 1, reports, resp)
+
+	// receiving heartbeats
+	req = newStoreHeartbeat(4, reports[4])
+	resp = &pdpb.StoreHeartbeatResponse{}
+	recoveryController.HandleStoreHeartbeat(req, resp)
+	re.Nil(resp.RecoveryPlan)
+	req = newStoreHeartbeat(1, reports[1])
+	resp = &pdpb.StoreHeartbeatResponse{}
+	recoveryController.HandleStoreHeartbeat(req, resp)
+	// enter demote failed voter stage
+	re.Equal(recoveryController.GetStage(), demoteFailedVoter)
+	re.NotNil(resp.RecoveryPlan)
+	re.Equal(len(resp.RecoveryPlan.Tombstones), 1)
+	applyRecoveryPlan(re, 1, reports, resp)
+
+	req = newStoreHeartbeat(4, reports[4])
+	resp = &pdpb.StoreHeartbeatResponse{}
+	recoveryController.HandleStoreHeartbeat(req, resp)
+	re.NotNil(resp.RecoveryPlan)
+	re.Equal(len(resp.RecoveryPlan.Demotes), 1)
+	applyRecoveryPlan(re, 1, reports, resp)
+}

--- a/server/cluster/unsafe_recovery_controller_test.go
+++ b/server/cluster/unsafe_recovery_controller_test.go
@@ -1055,10 +1055,8 @@ func TestTimeout(t *testing.T) {
 
 	time.Sleep(time.Second)
 	req := newStoreHeartbeat(1, nil)
+	req.StoreReport = &pdpb.StoreReport{Step: 1}
 	resp := &pdpb.StoreHeartbeatResponse{}
-	recoveryController.HandleStoreHeartbeat(req, resp)
-	re.Equal(exitForceLeader, recoveryController.GetStage())
-	req.StoreReport = &pdpb.StoreReport{Step: 2}
 	recoveryController.HandleStoreHeartbeat(req, resp)
 	re.Equal(failed, recoveryController.GetStage())
 }

